cpp/src/lists/contains.cu

```diff
@@ -0,0 +1,266 @@
+      return thrust::none_of(thrust::seq,
+                             counting_iter(0),
+                             counting_iter(list.size()),
+                             [&list] __device__(auto const& i) { return list.is_null(i); });
+    });
+}
+
+std::pair<rmm::device_buffer, size_type> construct_null_mask(
+  cudf::detail::lists_column_device_view const& d_lists,
+  cudf::column_device_view const& d_skeys,
+  bool input_column_has_nulls,
+  bool skeys_column_has_nulls,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  using namespace cudf;
+  using namespace cudf::detail;
+
+  if (!skeys_column_has_nulls && !input_column_has_nulls) {
+    return std::make_pair(rmm::device_buffer{0, stream, mr}, size_type{0});
+  }
+
+  return cudf::detail::valid_if(counting_iter(0),
+                                counting_iter(d_lists.size()),
+                                [d_lists, d_skeys] __device__(auto const& row_index) {
+                                  if (d_skeys.is_null(row_index)) { return false; }
+
+                                  auto list = cudf::list_device_view(d_lists, row_index);
+
+                                  if (list.is_null()) { return false; }
+
+                                  return thrust::none_of(
+                                    thrust::seq,
+                                    counting_iter(0),
+                                    counting_iter(list.size()),
+                                    [&list] __device__(auto const& i) { return list.is_null(i); });
+                                });
+}
+
+struct lookup_functor {
+  template <typename T, typename... Args>
+  std::enable_if_t<!cudf::is_numeric<T>() && !std::is_same<T, cudf::string_view>::value, void>
+  operator()(Args&&...) const
+  {
+    CUDF_FAIL("lists::contains() is only supported on numeric types and strings.");
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_numeric<T>() || std::is_same<T, cudf::string_view>::value, void>
+  operator()(cudf::detail::lists_column_device_view const& d_lists,
+             cudf::scalar const& skey,
+             cudf::mutable_column_device_view output_bools,
+             rmm::cuda_stream_view stream,
+             rmm::mr::device_memory_resource* mr) const
+  {
+    assert(skey.is_valid() && "skey should have been checked for nulls by this point.");
+
+    auto h_scalar = static_cast<cudf::scalar_type_t<T> const&>(skey);
+    auto d_scalar = cudf::get_scalar_device_view(h_scalar);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      counting_iter(0),
+                      counting_iter(d_lists.size()),
+                      output_bools.data<bool>(),
+                      [d_lists, d_scalar] __device__(auto row_index) {
+                        auto list = cudf::list_device_view(d_lists, row_index);
+                        if (list.is_null()) { return false; }
+                        for (size_type i{0}; i < list.size(); ++i) {
```

Comment:
Could this for loop be replaced with a sequential `find_if` using a `pair_iterator`?

Reply:
I'm afraid `cudf::list_device_view` does not yet have iterators.

Reply:
Ok, I've added a `pair_iterator` to `list_device_view`, and tightened up that loop.